openlibrary/plugins/openlibrary/js/lazy-thing-preview.js

```diff
@@ -0,0 +1,134 @@
+// @ts-check
+import debounce from 'lodash/debounce';
+
+/**
+ * Responds to HTML like:
+ * ```html
+ *  <div
+ *      class="lazy-thing-preview"
+ *      data-key="/works/OL123W"
+ *      data-render-fn="$render_fn_name"
+ * ></div>
+ * ```
+ *
+ * And then calls `$render_fn_name` with the response from search engine.
+ * Bundles/dedupes requests throughout the page and does other smart things.
+ * Someday: Only load if in viewport.
+ * Currently only works with works/editions.
+ */
+export class LazyThingPreview {
+    constructor() {
+        /** @type {Array<{key: string, render_fn: Function}>} */
+        this.queue = [];
+        /** @type {Object<string, object>} */
+        this.cache = {};
+
+        this.renderDebounced = debounce(this.render.bind(this), 100);
+    }
+
+    init() {
+        $('.lazy-thing-preview').each((i, el) => {
+            this.push({
+                key: el.dataset.key,
+                render_fn_name: el.dataset.renderFn,
+            });
+        });
+    }
+
+    /**
+     * @param {{key: string, render_fn_name: string}} arg0
+     */
+    push({key, render_fn_name}) {
+        const render_fn = window[render_fn_name];
+        if (this.cache[key]) {
+            this.renderKey(key, render_fn, this.cache[key]);
+        } else {
+            this.queue.push({key, render_fn});
+            this.renderDebounced();
+        }
+    }
+
+    /**
+     * @param {string} key
+     * @param {Function} render_fn
+     * @param {object} book
+     */
+    renderKey(key, render_fn, book) {
+        const $el = $(`.lazy-thing-preview[data-key="${key}"]`);
+        $el.html(render_fn(book));
+    }
+
+    /**
+     * @param {string[]} keys
+     * @returns {Promise<object[]>}
+     */
+    async getThings(keys) {
+        const workKeys = keys.filter(key => key.startsWith('/works/'));
+        const editionKeys = keys.filter(key => key.startsWith('/books/'));
+        const authorKeys = keys.filter(key => key.startsWith('/authors/'));
+        const fields = 'key,type,cover_i,first_publish_year,author_name,title,subtitle,edition_count,editions';
+        let docs = [];
+        if (workKeys.length) {
+            const resp = await fetch(`/search.json?${new URLSearchParams({
+                q: `key:(${workKeys.join(' OR ')})`,
+                fields,
+                limit: '100',
```

Comment:
NB: in current setup, previews will break for lists if > 100 works. @cdrini recommends possibly lazy loading previews in viewport.